pkg/apis/kubedirector/v1beta1/kubedirectorcluster_types.go

```diff
@@ -65,6 +65,7 @@ type KubeDirectorClusterStatus struct {
 	LastNodeID              int64            `json:"lastNodeID"`
 	Roles                   []RoleStatus     `json:"roles"`
 	LastConnectionHash      string           `json:"lastConnectionHash"`
+	UpgradedRoles           map[string]bool  `json:"upgradedroles,omitempty"`
```

Comment:
I need to read/comment more about how this info is used, but a few top level comments:

* Rather than sticking the info up here, I would recommend to put it as a new status field (or just a boolean flag perhaps) in the RoleStatus object, so that each role's status reflects whether it is currently being upgraded. It makes it a little more faffy for our code to write and read the correct RoleStatus object based on the role name, but we have utility functions to take care of that.

* Regardless of where we add the status info, we do also need to correspondingly update the status stanza in the kdcluster CRD.

* Shouldn't we clear/reset this info once the role upgrade has completed? Or more broadly speaking, how is this handled for multiple upgrades in sequence?

* If this info is meant to reflect whether a role is currently being upgraded (and hasn't finished yet), then its name will probably need to be different, e.g. using the term "upgrading" rather than "upgraded".

Reply:
@joel-bluedata , I decided to try add `UpgradingMembers` map field to `RoleStatus` instead of `UpgradeRoles` map. It should be filled on the `syncRoles()` stage as `memberName -> <new role image tag>` key-value pairs, before patching the statefulset. Then, on the `syncMembers()` stage, it will be sequentially cleaned, pod by pod, when we will succesfully notify the current pod container to upgrade. 

Sure, I could made it just as `[]string`, but I prefer to work with maps than with string arrays.
In other option, it could be changed to the boolean `Upgrading` field of `MemberStatus` struct. `RoleStatus` struct contains the `Members []MembersStatus` field, so it is possible to iterate this array twice - at enabling `Upgrading` of each member to true and then at disabling to false, when upgrading notification succeded.